collect_app/src/test/java/org/odk/collect/android/preferences/AdminPasswordDialogFragmentTest.java

```diff
@@ -0,0 +1,108 @@
+
+@RunWith(AndroidJUnit4.class)
+@LooperMode(PAUSED)
+public class AdminPasswordDialogFragmentTest {
+
+    @Before
+    public void setup() {
+        RobolectricHelpers.overrideAppDependencyModule(new AppDependencyModule() {
+            @Override
+            public AdminPasswordProvider providesAdminPasswordProvider() {
+                return new StubAdminPasswordProvider();
+            }
+        });
+    }
+
+    @Test
+    public void enteringPassword_andClickingOK_callsOnCorrectAdminPasswordWithAction() {
+        TestActivityScenario<MockAdminPasswordDialogCallback> activityScenario = TestActivityScenario.launch(MockAdminPasswordDialogCallback.class);
+        activityScenario.onActivity(activity -> {
+            AdminPasswordDialogFragment fragment = new AdminPasswordDialogFragment();
+            fragment.setAction(STORAGE_MIGRATION);
+            fragment.show(activity.getSupportFragmentManager(), "tag");
+            shadowOf(getMainLooper()).idle();
+
+            fragment.getInput().setText("password");
+            ((AlertDialog) fragment.getDialog()).getButton(AlertDialog.BUTTON_POSITIVE).performClick();
+            shadowOf(getMainLooper()).idle();
+
+            assertThat(activity.onCorrectAdminPasswordCalledWith, equalTo(STORAGE_MIGRATION));
+        });
+    }
+
+    @Test
+    public void afterRecreating_enteringPassword_andClickingOK_callsOnCorrectAdminPasswordWithAction() {
+        TestActivityScenario<MockAdminPasswordDialogCallback> activityScenario = TestActivityScenario.launch(MockAdminPasswordDialogCallback.class);
+        activityScenario.onActivity(activity -> {
+            AdminPasswordDialogFragment fragment = new AdminPasswordDialogFragment();
+            fragment.setAction(STORAGE_MIGRATION);
+            fragment.show(activity.getSupportFragmentManager(), "tag");
+            shadowOf(getMainLooper()).idle();
+        });
+
+        activityScenario.recreate();
+        activityScenario.onActivity(activity -> {
+            AdminPasswordDialogFragment fragment = (AdminPasswordDialogFragment) activity.getSupportFragmentManager().findFragmentByTag("tag");
+
+            fragment.getInput().setText("password");
+            ((AlertDialog) fragment.getDialog()).getButton(AlertDialog.BUTTON_POSITIVE).performClick();
+            shadowOf(getMainLooper()).idle();
+
+            assertThat(activity.onCorrectAdminPasswordCalledWith, equalTo(STORAGE_MIGRATION));
+        });
+    }
+
+    private static class StubAdminPasswordProvider extends AdminPasswordProvider {
+
+        StubAdminPasswordProvider() {
+            super(null);
+        }
+
+        @Override
+        public boolean isAdminPasswordSet() {
+            return true;
+        }
+
+        @Override
+        public String getAdminPassword() {
+            return "password";
+        }
+    }
+
+    private static class MockAdminPasswordDialogCallback extends FragmentActivity implements AdminPasswordDialogFragment.AdminPasswordDialogCallback  {
```

Comment:
This feels more like a spy to me since it's keeping extra state and verifying it rather than holding an expectation of what it's supposed to receive. Not that important but would be good to be somewhat aligned. I'm referring to https://martinfowler.com/articles/mocksArentStubs.html.

Reply:
Ach my fav article. Hoisted with my own petard! Yeah you're dead on. In my head was using for "verification" but you're right that it's doing "recording". I'll fix this!

Reply:
> Hoisted with my own petard!

What a lovely expression. 💥 @grzesiek2010 do you know this one?!